### PR TITLE
[VPlan] Replace VPRecipeOrVPValue with VP2VP recipe simplification.

### PR DESCRIPTION
**Pull request by fhahn**
Move simplification of VPBlendRecipes from early VPlan construction to VPlan-to-VPlan based recipe simplification. This simplifies initial construction.

Note that some in-loop reduction tests are failing at the moment, due to the reduction predicate being created after the reduction recipe. I will provide a patch for that soon.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -9079,6 +9043,9 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
         LinkVPBB->insert(FMulRecipe, CurrentLink->getIterator());
         VecOp = FMulRecipe;
       } else {
+        if (PhiR->isInLoop() && isa<VPBlendRecipe>(CurrentLink))
+          continue;
+
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9048]: Replace the blend here by its non-header-phi operand (possibly after the loop if needed, and/or updating PreviousLink), as mentioned below?

**Review by ayalz — COMMENTED** [review]
Nice cleanup, thanks for pursuing!

Wonder if `tryTo*()`'s should return more specific recipes than `VPRecipeBase*`, as in `tryToBlend()`.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -8966,10 +8930,10 @@ VPlanPtr LoopVectorizationPlanner::buildVPlan(VFRange &Range) {
...
 
 // Adjust the recipes for reductions. For in-loop reductions the chain of
-// instructions leading from the loop exit instr to the phi need to be converted
-// to reductions, with one operand being vector and the other being the scalar
-// reduction chain. For other reductions, a select is introduced between the phi
-// and live-out recipes when folding the tail.
+// instructions leading from the loop exit instr to the phi need to be
+// converted to reductions, with one operand being vector and the other
+// being the scalar reduction chain. For other reductions, a select is
+// introduced between the phi and live-out recipes when folding the tail.
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: undone, thanks!

**Review by fhahn — COMMENTED** [review]
Thanks for taking a look, comments should be addressed!


>  Wonder if tryTo*()'s should return more specific recipes than VPRecipeBase*, as in tryToBlend().


That would probably be good now! Should I adjust the current patch or do as follow-up?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -8999,6 +8963,18 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
         LinkVPBB->insert(FMulRecipe, CurrentLink->getIterator());
         VecOp = FMulRecipe;
       } else {
+        auto *Blend = dyn_cast<VPBlendRecipe>(CurrentLink);
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8966]: Add folding of inloop blend to documentation above.

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 4]
```diff
@@ -88,8 +86,8 @@ class VPRecipeBuilder {
   /// or a new VPBlendRecipe otherwise. Currently all such phi nodes are turned
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:86]: Comment needs updating.

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 5]
```diff
@@ -116,12 +111,11 @@ class VPRecipeBuilder {
 
   /// Check if an existing VPValue can be used for \p Instr or a recipe can be
   /// create for \p I withing the given VF \p Range. If an existing VPValue can
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:113]: Comment needs updating. (Plus fixing "withing" typo).

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 6]
```diff
@@ -806,6 +806,19 @@ static unsigned getOpcodeForRecipe(VPRecipeBase &R) {
...
+    bool AllEqual = true;
+    for (unsigned I = 1; I != Blend->getNumIncomingValues(); ++I)
+      AllEqual &= Blend->getIncomingValue(0) == Blend->getIncomingValue(I);
+    if (AllEqual) {
+      Blend->replaceAllUsesWith(Blend->getIncomingValue(0));
+      Blend->eraseFromParent();
+      return;
+    }
+    return;
+  }
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:820]: Can be simplified?
```suggestion
  if (auto *Blend = dyn_cast<VPBlendRecipe>(&R)) {
    for (unsigned I = 1; I != Blend->getNumIncomingValues(); ++I)
      if (Blend->getIncomingValue(0) != Blend->getIncomingValue(I))
        return;
    Blend->replaceAllUsesWith(Blend->getIncomingValue(0));
    Blend->eraseFromParent();
    return;
  }
```

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 7]
```diff
@@ -8999,6 +8963,18 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
...
+        if (PhiR->isInLoop() && Blend) {
+          assert(Blend->getNumIncomingValues() == 2);
+          assert(any_of(Blend->operands(),
+                        [PhiR](VPValue *Op) { return Op == PhiR; }));
+          if (Blend->getIncomingValue(0) == PhiR)
+            Blend->replaceAllUsesWith(Blend->getIncomingValue(1));
+          else
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: done, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 6 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Simplified thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 8]
```diff
@@ -9027,7 +8994,8 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
     // the phi until LoopExitValue. We keep track of the previous item
     // (PreviousLink) to tell which of the two operands of a Link will remain
     // scalar and which will be reduced. For minmax by select(cmp), Link will be
-    // the select instructions.
+    // the select instructions. Blend recipes will get folded to their non-phi
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8997]:
```suggestion
    // the select instructions. Blend recipes of in-loop reduction phi's will get folded to their non-phi
```

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 9]
```diff
@@ -9058,6 +9026,20 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
...
+        auto *Blend = dyn_cast<VPBlendRecipe>(CurrentLink);
+        if (PhiR->isInLoop() && Blend) {
+          assert(Blend->getNumIncomingValues() == 2 &&
+                 "Blend must have 2 incoming values");
+          if (Blend->getIncomingValue(0) == PhiR)
+            Blend->replaceAllUsesWith(Blend->getIncomingValue(1));
+          else {
+            Blend->replaceAllUsesWith(Blend->getIncomingValue(0));
+            assert(Blend->getIncomingValue(1) == PhiR &&
+                   "PhiR must be an operand of the blend");
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9038]: nit: may be slightly more consistent to have the assert first.

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 10]
```diff
@@ -100,11 +99,8 @@ class VPRecipeBuilder {
   /// Check if \p I has an opcode that can be widened and return a VPWidenRecipe
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:99]: (Indeed back to returning VPWidenRecipe as documented...)

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 11]
```diff
@@ -827,6 +827,16 @@ static unsigned getOpcodeForRecipe(VPRecipeBase &R) {
 
 /// Try to simplify recipe \p R.
 static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+  // Try to remove redundant blend recipes.
+  if (auto *Blend = dyn_cast<VPBlendRecipe>(&R)) {
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:831]: nit: may be worth defining and reusing `IncomingValue0 = Blend->getIncomingValue(0);`

**Review by ayalz — APPROVED** [review]
Looks good to me, thanks for cleaning this up!
Adding a couple of minor nits.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 11 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:831]: Done, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 9 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9052]: Done, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 8 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9011]: Done, thanks!

**Pull request merged**